command/operator_diagnose.go

```diff
@@ -140,6 +152,61 @@ func (c *OperatorDiagnoseCommand) RunWithParsedFlags() int {
 		return 1
 	}
 
+	// Check Listener Information
+	// TODO: Run Diagnose checks on the actual net.Listeners
+
+	disableClustering := config.HAStorage.DisableClustering
+	infoKeys := make([]string, 0, 10)
+	info := make(map[string]string)
+	status, errMsg, lns, _ := server.InitListeners(config, disableClustering, &infoKeys, &info)
+
+	if status != 0 {
+		c.UI.Output("Error parsing listener configuration.")
+		c.UI.Error(errMsg.Error())
+		return 1
+	}
+
+	// Make sure we close all listeners from this point on
+	listenerCloseFunc := func() {
+		for _, ln := range lns {
+			ln.Listener.Close()
+		}
+	}
+
+	defer c.cleanupGuard.Do(listenerCloseFunc)
+
+	sanitizedListeners := make([]listenerutil.Listener, 0, len(config.Listeners))
```

Comment:
how helpful is it to specify a capacity here? it appears that the required capacity can be lower than `len(config.Listeners)` (if `ln.Config.TLSDisable`, for example)

Reply:
Same as above -- we should avoid using dynamically sized arrays where possible. In this case, I'm willing to preallocate the listener length, even if we don't necessarily use all the space. It would be more costly I think to append individually.

Reply:
is this a pretty costly operation?

Reply:
It doubles the array size every time and copies all elements into a new array. https://blog.golang.org/slices-intro for more info!

Reply:
i meant for this code specifically. like how many listeners are typical?

Reply:
I think either way it's not too big a deal. In the case of 3 listeners, we'd end up with an array of 4 with dynamic sizing, versus 3 in the exact case. In the case of 3 listeners with 1 tls_disable listener, we'd end up with 3 in the exact case versus 2 in the dynamic resizing case. 

So in this particular case where the size of the number of listeners is small, I'd say it's more of an idiomatic/personal preference matter.

Reply:
okay cool on the personal preference bit. i was just curious why that choice was made